### DISCUSSION TITLE
Loading finetuned RT-DETR model

### DISCUSSION THREAD

**Discussion by arabinda-iitbbs** [ultralytics/ultralytics#13751] [Q&A]
I've successfully fine-tuned RT-DETR model, and I have the best weights saved at "runs/detect/train3/weights/best.pt". Now, i want to load these weights to validate model and perform inference. I also want to print my results value like mAP of my test data. can you guys please guide me in doing this? please @glenn-jocher @pderrenger

**Comment by pderrenger** [comment]
Hello!

Great to hear that you've successfully fine-tuned the RT-DETR model! To load your saved weights and validate the model or perform inference, you can follow these steps:

### Load and Validate the Model
To load your fine-tuned model and validate it, you can use the following code:

```python
from ultralytics import RTDETR

# Load your fine-tuned model
model = RTDETR("runs/detect/train3/weights/best.pt")

# Validate the model
results = model.val(data="path/to/your/test_data.yaml")

# Print the mAP results
print(f"mAP50-95: {results.box.map:.4f}")
print(f"mAP50: {results.box.map50:.4f}")
print(f"mAP75: {results.box.map75:.4f}")
```

### Perform Inference
For inference on new images, you can use:

```python
# Perform inference on an image
results = model("path/to/your/image.jpg")

# Display results
results.show()
```

### Using CLI
Alternatively, you can use the command line interface:

```bash
# Validate the model
yolo val model=runs/detect/train3/weights/best.pt data=path/to/your/test_data.yaml

# Perform inference
yolo predict model=runs/detect/train3/weights/best.pt source=path/to/your/image.jpg
```

### Ensure Latest Versions
Please ensure you are using the latest versions of `torch` and `ultralytics` packages. You can upgrade them using:

```bash
pip install --upgrade torch ultralytics
```

If you encounter any issues, please provide a minimum reproducible example as outlined [here](https://docs.ultralytics.com/help/minimum_reproducible_example). This will help us investigate and resolve the issue more effectively.

Feel free to reach out if you have any more questions. Happy coding! 😊

**Reply by arabinda-iitbbs** [reply]
Thank you for the reply, @pderrenger. I have one question: What is test_data.yaml? My data.yaml file structure is like this: I have included all the train, val and test folder path information in the "data. yaml" file. 

"train: F:/arabinda/rt-detr/Augmented_Dataset_COCO/train
val: F:/arabinda/rt-detr/Augmented_Dataset_COCO/valid
test: F:/arabinda/rt-detr/Augmented_Dataset_COCO/test
nc: 13
names: ['Multi-axle truck', 'auto', 'bike', 'bus', 'car', 'cycle', 'lcv', 'minivan', 'pick up', 'rickshaw', 'tractor', 'tractor with trailer', 'truck']."
 Inside each train, val and test, the corresponding JSON file contains the annotation information.
 Can you tell me whether my yaml structure is correct or not?
 If it is correct, then in my case, to validate the model, what should I write for validating the model code "results = model. val(data=" path/to/your/test_data.yaml")" in place of "data=" path/to/your/test_data.yaml."

**Reply by pderrenger** [reply]
Hello @arabinda-iitbbs,

Thank you for your detailed question! Your `data.yaml` structure looks correct for specifying the paths to your training, validation, and test datasets. Given this structure, you can use the same `data.yaml` file for both training and validation.

To validate your model using the provided `data.yaml`, you can simply point to it directly. Here's how you can do it:

### Validate the Model
```python
from ultralytics import RTDETR

# Load your fine-tuned model
model = RTDETR("runs/detect/train3/weights/best.pt")

# Validate the model using your data.yaml
results = model.val(data="F:/arabinda/rt-detr/Augmented_Dataset_COCO/data.yaml")

# Print the mAP results
print(f"mAP50-95: {results.box.map:.4f}")
print(f"mAP50: {results.box.map50:.4f}")
print(f"mAP75: {results.box.map75:.4f}")
```

### Using CLI
Alternatively, you can validate the model using the CLI:

```bash
yolo val model=runs/detect/train3/weights/best.pt data=F:/arabinda/rt-detr/Augmented_Dataset_COCO/data.yaml
```

This will ensure that your model is validated using the test dataset specified in your `data.yaml`.

If you encounter any issues, please ensure you are using the latest versions of `torch` and `ultralytics` packages. You can upgrade them using:

```bash
pip install --upgrade torch ultralytics
```

If the issue persists, please provide a minimum reproducible example as outlined [here](https://docs.ultralytics.com/help/minimum_reproducible_example). This will help us investigate and resolve the issue more effectively.

Feel free to reach out if you have any more questions. Happy coding! 😊

**Reply by arabinda-iitbbs** [reply]
Thank you @pderrenger. I have one more question I want to ask. Can we also train our RT-DETR using the YOLOv8 data format?

**Reply by pderrenger** [reply]
Hello @arabinda-iitbbs,

Absolutely, you can train your RT-DETR model using the YOLOv8 data format. The YOLOv8 data format is quite flexible and can be used with various models, including RT-DETR.

Here's a quick guide on how to set up your data for training RT-DETR using the YOLOv8 format:

1. **Data Structure**: Ensure your `data.yaml` file is correctly structured, as you have already done:
    ```yaml
    train: F:/arabinda/rt-detr/Augmented_Dataset_COCO/train
    val: F:/arabinda/rt-detr/Augmented_Dataset_COCO/valid
    test: F:/arabinda/rt-detr/Augmented_Dataset_COCO/test
    nc: 13
    names: ['Multi-axle truck', 'auto', 'bike', 'bus', 'car', 'cycle', 'lcv', 'minivan', 'pick up', 'rickshaw', 'tractor', 'tractor with trailer', 'truck']
    ```

2. **Annotations**: Ensure that your annotations are in the correct format (e.g., COCO JSON format) within the specified directories.

3. **Training**: Use the following code to train your RT-DETR model:
    ```python
    from ultralytics import RTDETR

    # Load the model configuration
    model = RTDETR("path/to/your/model.yaml")

    # Train the model
    model.train(data="F:/arabinda/rt-detr/Augmented_Dataset_COCO/data.yaml", epochs=50, imgsz=640)
    ```

4. **Validation and Inference**: You can validate and perform inference as previously discussed.

If you encounter any issues, please ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:
```bash
pip install --upgrade torch ultralytics
```

If you need further assistance, please provide a minimum reproducible example as outlined [here](https://docs.ultralytics.com/help/minimum_reproducible_example). This will help us investigate and resolve the issue more effectively.

Feel free to reach out if you have any more questions. Happy coding! 😊

**Reply by arabinda-iitbbs** [reply]
Thank you, @pderrenger, for your help. Can we integrate the DeepSORT algorithm to track the vehicles after the detection is done using RT-DETR?

**Reply by pderrenger** [reply]
Hello @arabinda-iitbbs,

You're welcome! Yes, you can integrate the DeepSORT algorithm for tracking vehicles after detection with RT-DETR. DeepSORT is a popular tracking algorithm that works well with object detectors like RT-DETR.

Here's a concise guide on how to achieve this integration:

### Step-by-Step Integration

1. **Install DeepSORT**: Ensure you have the DeepSORT dependencies installed. You can find the DeepSORT repository [here](https://github.com/nwojke/deep_sort).

2. **Modify Your Inference Script**: After detecting objects with RT-DETR, pass the detection results to DeepSORT for tracking.

3. **Example Code**:
    ```python
    from ultralytics import RTDETR
    from deep_sort_realtime.deepsort_tracker import DeepSort

    # Load your fine-tuned RT-DETR model
    model = RTDETR("runs/detect/train3/weights/best.pt")

    # Initialize DeepSORT
    tracker = DeepSort(max_age=30, n_init=3, nms_max_overlap=1.0, max_cosine_distance=0.2)

    # Perform inference on an image
    results = model("path/to/your/image.jpg")

    # Extract bounding boxes and confidence scores
    bboxes = results[0].boxes.xyxy.cpu().numpy()
    scores = results[0].boxes.conf.cpu().numpy()

    # Update tracker with detections
    tracks = tracker.update_tracks(bboxes, scores, frame_id=0)

    # Display tracking results
    for track in tracks:
        print(f"Track ID: {track.track_id}, BBox: {track.to_tlbr()}")
    ```

4. **Ensure Compatibility**: Make sure the bounding box format and other parameters are compatible between RT-DETR and DeepSORT.

### Additional Resources
- [DeepSORT Documentation](https://github.com/nwojke/deep_sort)
- [Ultralytics RT-DETR Documentation](https://docs.ultralytics.com/models/rtdetr/)

If you encounter any issues, please ensure you are using the latest versions of the packages. If the problem persists, providing a minimum reproducible example can greatly help in diagnosing the issue. You can find guidelines for creating one [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions. Happy coding! 😊

**Comment by maxcodesfast** [comment]
Hi,
I run into a problem while converting a fine tuned rt-detr-l to tensorRT. The error is: RuntimeError: 

failed to load ONNX file: /media/shiptec/Seagate/Masterarbeit/results_training/results_MA/LakeCn_merged_2_detr_l/train_train/wAug_nFreeze_nRect_rVal/weights/best.onnx 

I use this code:

from ultralytics import RTDETR
# Export model
model = RTDETR('path_to_model/weights/best.pt')  # load a custom trained model
model.export(format='engine', imgsz=(416,640))  

The code is running on a jetson orin 64gb with the following software versions:
ultralytics: 8.2.2
torch: 2.1.0a0+41361538.nv23.06
torchvision: 0.16.2+c6f3977
numpy: 1.23.5
trt: 8.5.2.2

Can someone help me with that?

**Reply by pderrenger** [reply]
Hello,

Thank you for reaching out! It looks like you're encountering an issue while converting your fine-tuned RT-DETR model to TensorRT format. Let's address this step-by-step.

### Loading and Validating the Model
To load your fine-tuned RT-DETR model and validate it, you can use the following code snippet:

```python
from ultralytics import RTDETR

# Load the fine-tuned model
model = RTDETR('runs/detect/train3/weights/best.pt')

# Validate the model
metrics = model.val(data='path_to_your_validation_data.yaml')
print(f"mAP50-95: {metrics.box.map:.4f}")
```

### Performing Inference
To perform inference with your fine-tuned model, you can use:

```python
# Perform inference
results = model('path_to_your_test_image_or_directory')
results.show()  # Display results
results.save(save_dir='path_to_save_results')  # Save results
```

### Exporting to TensorRT
Regarding the error you're encountering during the export to TensorRT, it seems there might be an issue with the ONNX file generation or compatibility. Here are a few steps to troubleshoot:

1. **Verify ONNX Export**: Ensure that the ONNX export works correctly before converting to TensorRT.
    ```python
    model.export(format='onnx', imgsz=(416, 640))
    ```

2. **Check ONNX File**: Verify the generated ONNX file using an ONNX runtime or viewer to ensure it is correctly formatted.

3. **Update Packages**: Make sure you are using the latest versions of all relevant packages. You can update Ultralytics and other dependencies using:
    ```bash
    pip install --upgrade ultralytics
    pip install --upgrade torch torchvision
    ```

4. **TensorRT Compatibility**: Ensure that the TensorRT version you are using is compatible with the versions of PyTorch and ONNX you have installed.

### Example Code for Exporting to TensorRT
Here is an example code snippet for exporting to TensorRT:

```python
from ultralytics import RTDETR

# Load the fine-tuned model
model = RTDETR('runs/detect/train3/weights/best.pt')

# Export to TensorRT
model.export(format='engine', imgsz=(416, 640))
```

If the issue persists, please provide a minimum reproducible example and verify that the issue occurs with the latest versions of the packages. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions or need further assistance. 😊
